pennylane_lightning/src/simulator/StateVectorManagedCPU.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include "BitUtil.hpp"
+#include "CPUMemoryModel.hpp"
+#include "Gates.hpp"
+#include "KernelType.hpp"
+#include "Memory.hpp"
+#include "StateVectorBase.hpp"
+#include "StateVectorCPU.hpp"
+#include "Threading.hpp"
+#include "Util.hpp"
+
+namespace Pennylane {
+
+/**
+ * @brief StateVector class where data resides in CPU memory. Memory ownership
+ * resides within class.
+ *
+ * @tparam PrecisionT
+ */
+template <class PrecisionT = double>
+class StateVectorManagedCPU
+    : public StateVectorCPU<PrecisionT, StateVectorManagedCPU<PrecisionT>> {
+  public:
+    using ComplexPrecisionT = std::complex<PrecisionT>;
+
+  private:
+    using BaseType = StateVectorCPU<PrecisionT, StateVectorManagedCPU>;
+
+    std::vector<ComplexPrecisionT, Util::AlignedAllocator<ComplexPrecisionT>>
+        data_;
+
+  public:
+    /**
+     * @brief Create a new statevector
+     *
+     * @param num_qubits Number of qubits
+     * @param threading Threading option the statevector to use
+     * @param memory_model Memory model the statevector will use
+     */
+    explicit StateVectorManagedCPU(
+        size_t num_qubits, Threading threading = Threading::SingleThread,
+        CPUMemoryModel memory_model = bestCPUMemoryModel())
+        : BaseType{num_qubits, threading, memory_model},
+          data_{Util::exp2(num_qubits), ComplexPrecisionT{0.0, 0.0},
+                getAllocator<ComplexPrecisionT>(this->memory_model_)} {
+        data_[0] = {1, 0};
+    }
+
+    /**
+     * @brief Construct a statevector from another statevector
+     *
+     * @param other Another statevector to construct the statevector from
```

Comment:
Please add some description for template params here and other places in this file...

Reply:
Added!